### PR TITLE
Round robin during spread scheduling

### PR DESCRIPTION
**Pull request by jjyao**
## Why are these changes needed?
This PR does round robin over all the nodes instead of always starting from the beginning of the node list for spread scheduling.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
Nice change!

**Comment by jjyao** [comment]
pipelined_ingestion_1500_gb_15_windows succeeded without prefix hack: https://buildkite.com/ray-project/periodic-ci/builds/1902#_

**Comment by ericl** [comment]
How's the before/after time? And time without spread hint?

**Comment by jjyao** [comment]
Without prefix, without spread scheduling strategy: Forever (hang)
With prefix: ~33m
With spread scheduling strategy: ~47m

@scv119 can confirm this.

**Comment by ericl** [comment]
Any idea why it's still 50% slower? Btw, .stats() can show you the node
distribution though it might not be complete for shuffle.

**Comment by jjyao** [comment]
I'm trying to figure out the difference now. This is a good validation use case for dataset stats and scheduler observability projects.

File: python/ray/data/impl/shuffle.py
Comment (ericl) [inline python/ray/data/impl/shuffle.py:34]: Use the "SPREAD" syntax?

Comment (jjyao) [reply python/ray/data/impl/shuffle.py:34]: SPREAD_SCHEDULING_STRATEGY is a variable for "SPREAD". Do you mean use "SPREAD" literal directly?

File: src/ray/raylet/scheduling/scheduling_policy.cc
Comment (rkooo567) [inline src/ray/raylet/scheduling/scheduling_policy.cc:134]: nit, but why don't we just set
```
size_t round_index = scheduling_next_index_
```

and change this code to 
```
      if (spread_threshold == 0) {
        // SPREAD
        scheduling_next_index_ = ((round_index + 1) % round.size());
      } else {
         // PACK
         scheduling_next_index_ = 0;
     }
```

File: src/ray/raylet/scheduling/scheduling_policy.cc
Comment (rkooo567) [inline src/ray/raylet/scheduling/scheduling_policy.cc:49]: Is it safe to remove this? Kind of wondering why we have this in the first place lol? It seems a little weird

File: src/ray/raylet/scheduling/scheduling_policy.cc
Comment (rkooo567) [inline src/ray/raylet/scheduling/scheduling_policy.cc:85]: nice change!

File: src/ray/raylet/scheduling/scheduling_policy_test.cc
Comment (rkooo567) [inline src/ray/raylet/scheduling/scheduling_policy_test.cc:351]: Can you write comment what the test does?

**Review by rkooo567 — COMMENTED** [review]
LGTM for cpp part. I will wait for Eric's approval since I didn't review the dataset part. 

Only one nit comment

File: python/ray/data/impl/shuffle.py
Comment (ericl) [reply python/ray/data/impl/shuffle.py:34]: Yeah, that's what we document in the API, so we should prefer using that. But not a strong preference

File: src/ray/raylet/scheduling/scheduling_policy.h
Comment (clarkzinzow) [inline src/ray/raylet/scheduling/scheduling_policy.h:75]: Hmm if we kept the node IDs in a sorted container (`std::set`, `absl::btree_set`, or `boost::flat_set`) that's properly updated upon node addition or removal, and we used that to keep `spread_scheduling_next_index_` updated, we could:
1. Keep `spread_scheduling_next_index_` accurate in the face of nodes being added and removed.
2. Obviate the need for a sort of all node IDs on the critical scheduling path. Since node membership changes will most likely be far less frequent than task scheduling, this could be a (small) scheduling performance win for large clusters.

Probably not worth the complexity for now.

File: python/ray/util/scheduling_strategies.py
Comment (clarkzinzow) [inline python/ray/util/scheduling_strategies.py:15]: 🙌

**Review by clarkzinzow — COMMENTED** [review]
Datasets side looks good, a few comments about the Ray side.

**Comment by jjyao** [comment]
Not sure if there is anything changed or last run was just noise: I ran `pipelined_ingestion_1500_gb_15_windows` three times with this PR and the time is almost the same now: 
https://buildkite.com/ray-project/periodic-ci/builds/2059#_
https://buildkite.com/ray-project/periodic-ci/builds/2057
https://buildkite.com/ray-project/periodic-ci/builds/2054

File: src/ray/raylet/scheduling/scheduling_policy.cc
Comment (jjyao) [reply src/ray/raylet/scheduling/scheduling_policy.cc:134]: As @clarkzinzow mentioned, if we do `scheduling_next_index_ = 0;`, a pack task will affect the round robin of the spread task.

File: src/ray/raylet/scheduling/scheduling_policy.cc
Comment (jjyao) [reply src/ray/raylet/scheduling/scheduling_policy.cc:49]: It's not removed. I just moved it to a different place.

File: src/ray/raylet/scheduling/scheduling_policy.cc
Comment (jjyao) [reply src/ray/raylet/scheduling/scheduling_policy.cc:68]: Yea, we can also maintain a sorted list of nodes instead of resorting every time. I'll leave this out for a future optimization.

**Review by rkooo567 — APPROVED** [review]

File: src/ray/raylet/scheduling/scheduling_policy.cc
Comment (scv119) [reply src/ray/raylet/scheduling/scheduling_policy.cc:134]: I think a lot of complexity comes from the mixture of pack and spread policy in the same function. wondering if it's easier if we separate those two.

Comment (jjyao) [reply src/ray/raylet/scheduling/scheduling_policy.cc:134]: They actually have almost the same implementation: the only difference is where we start the scan.

**Comment by jjyao** [comment]
It's ready to be merged. Perf number looks good: https://buildkite.com/ray-project/periodic-ci/builds/2086#_

**Review by ericl — APPROVED** [review]

**Comment by jjyao** [comment]
Oh, I think for latests runs, I ran the tests against master which still had the prefix code in `pipelined_training.py`. Let me revert the PR to fix `pipelined_ingestion_1500_gb_15_windows`.

File: src/ray/raylet/scheduling/scheduling_policy.cc
Comment (clarkzinzow) [reply src/ray/raylet/scheduling/scheduling_policy.cc:68]: Yep as I mentioned in [this comment](https://github.com/ray-project/ray/pull/19968#discussion_r769206622), a standard sorted container would have a few benefits but isn't worth the complexity at this point IMO. Not a high priority since this is most likely not even in the top 3 scalability/perf bottlenecks for our scalability envelope, but we should make sure that our large-cluster scalability test measures this scheduling slowdown so we can easily pinpoint it once it _does_ become the primary bottleneck.

**Pull request merged**
